src/components/multi-select/multi-select.jsx

```diff
@@ -0,0 +1,316 @@
+import React, {
+  createRef,
+  useEffect,
+  useImperativeHandle,
+  useRef,
+  useState,
+  forwardRef,
+} from 'react';
+import PropTypes from 'prop-types';
+import FormControl from '../form-control/form-control.jsx';
+import FormControlIcons from '../form-control-icons/form-control-icons.jsx';
+import Icon from '../icon/icon.jsx';
+import IconButton from '../icon-button/icon-button.jsx';
+import iconCaretDown from '../../svgs/caret-down.svg';
+import iconClear from '../../svgs/clear.svg';
+import Listbox from '../listbox/listbox.jsx';
+import ListOption from '../list-option/list-option.jsx';
+import Loader from '../loader/loader.jsx';
+import NoOptions from '../no-options/no-options.jsx';
+import TagList from '../tag-list/tag-list.jsx';
+import Tag from '../tag/tag.jsx';
+import useDropdownClose from '../../hooks/use-dropdown-close.js';
+import styles from './multi-select.css';
+
+const MultiSelect = forwardRef(function MultiSelect(props, ref) {
+  const [autocompleteValue, setAutocompleteValue] = useState('');
+  const autocompleteRef = useRef();
+  const backupRef = useRef();
+  const [expanded, setExpanded] = useState(false);
+  const selfRef = ref || backupRef;
+  const [validationMessage, setValidationMessage] = useState(props.validationMessage);
+  const [value, setValue] = useState(props.value || []);
+  const valueRefs = value.map(() => createRef());
+  const visibleOptions = getVisibleOptions();
+  const visibleOptionsRefs = visibleOptions.map(() => createRef());
+  const wrapperRef = useRef(null);
+
+  const ids = {
+    emptyMessage: `${props.id}-empty`,
+    label: `${props.id}-label`,
+    listbox: `${props.id}-listbox`,
+    textbox: `${props.id}-autocomplete`,
+  };
+
+  const classNames = {
+    container: styles.container,
+    formControlChildrenContainer: props.readOnly ? styles['form-control-children-container-readonly'] : styles['form-control-children-container'],
```

Comment:
I think this is more aptly named as something along the lines of "input" and the input below is more of a "combobox"? 

Or maybe this is: `combobox-container` (since it contains a combobox)? 

This does separate the classnames from the React and instead relies on layout and roles. 🤔  

What do you think?

Reply:
Yeah, that's an interesting viewpoint. Do we want the class names to reflect what is in JSX, or what is in the accessibility tree? I think I've tended towards what is in JSX because not all sets of elements that need styling have any relevant accessibility. Not sure that's a great reason to not mix the two, but I think it does make it simpler to grok.

Reply:
Additionally, I think renaming `formControlChildrenContainer` to `input` may be out-and-out confusing for anyone reading the code when they see that the literal `input` is not the `input` _className`. 😬